File: services/fleet-manager/app/manager.py
# /fleet-manager/manager.py
import os
import copy
from kubernetes import client, config
from kubernetes.client.rest import ApiException

from app.core.config import settings
# --- 1. LOAD CONFIGURATIONS ---

# --- 2. THE MANAGER (The "Kubernetes-Native" Brain) ---
class ServerManager:
    def __init__(self, game_templates, game_settings):
        try:
            config.load_incluster_config()
        except config.ConfigException:
            config.load_kube_config() # Fallback for local testing
        
        self.GAME_TEMPLATES = game_templates
        self.SETTINGS = game_settings
        self.core_v1 = client.CoreV1Api()
        self.custom_api = client.CustomObjectsApi()
        
        self.namespace = settings.NAMESPACE
        
        agones_settings = self.SETTINGS.get('agones', {})

        self.group = agones_settings.get('group', 'agones.dev')
        self.version = agones_settings.get('version', 'v1')
        self.plural = agones_settings.get('plural', 'gameservers')

    def _ensure_pvc_exists(self, logical_server_id: str):
        """Ensures the user has a permanent hard drive for this specific game."""
        pvc_name = f"world-pvc-{logical_server_id}"
        
        try:
            self.core_v1.read_namespaced_persistent_volume_claim(
                name=pvc_name, namespace=self.namespace
            )
        except ApiException as e:
            if e.status == 404:
                print(f"Provisioning new storage volume: {pvc_name}")
                pvc_manifest = client.V1PersistentVolumeClaim(
                    metadata=client.V1ObjectMeta(name=pvc_name),
                    spec=client.V1PersistentVolumeClaimSpec(
                        access_modes=["ReadWriteOnce"],
                        resources=client.V1ResourceRequirements(requests={"storage": "10Gi"})
                    )
                )
                self.core_v1.create_namespaced_persistent_volume_claim(
                    namespace=self.namespace, body=pvc_manifest
                )
            else:
                raise e
        return pvc_name

    def _build_manifest(self, game_id: str, user_id: str, config_data: dict, logical_server_id: str, sidecar_token: str):
        """Constructs the Agones GameServer YAML using the Template Registry."""
        # Determine mod state from config (default to vanilla)
        is_modded = config_data.get("is_modded", False)
        mod_state = "modded" if is_modded else "vanilla"
        
        # Fetch the blueprint
        blueprint = self.GAME_TEMPLATES.get(game_id, {}).get(mod_state)
        if not blueprint:
            raise ValueError(f"No template found for {game_id} ({mod_state})")

        # Provision/Fetch the player's storage drive
        pvc_name = self._ensure_pvc_exists(logical_server_id)
        
        # Merge user config with blueprint defaults
        final_env = copy.deepcopy(blueprint.get("env_defaults", []))
        for k, v in config_data.items():
            if k != "is_modded": # Don't pass our internal flag to the game
                final_env.append({"name": k, "value": str(v)})
        
        app_settings = self.SETTINGS.get('app', {})
        manager_api_url = settings.MANAGER_API_URL
        telemetry_api_url = settings.TELEMETRY_API_URL
        
        # 1. Grab the initContainers from the blueprint (if they exist)
        init_containers = copy.deepcopy(blueprint.get("initContainers", []))
        
        # 2. Inject the Mod URLs into the initContainer's environment
        mod_urls_string = config_data.get("mod_urls", "") # Your API passes this in
        
        for ic in init_containers:
            if ic["name"] == "mod-downloader":
                for env_var in ic.get("env", []):
                    if env_var["name"] == "MOD_URLS":
                        env_var["value"] = mod_urls_string

        # 3. Add initContainers to the pod spec
        # I have a side car image setup, just need to change the Avtivity-sidecar
        # replace image -- remove CMD -- remove volume sidecar-scripts-volume
        manifest_spec_template_spec = {
            "containers": [
                {
                    "name": "game-engine",
                    "image": blueprint["image"],
                    "env": final_env,
                    "volumeMounts": [{"name": "world-data", "mountPath": "/config"}]
                },
                {
                    "name": "activity-sidecar",
                    "image": "python:3.14.1-alpine3.23",
                    "command": ["python", "-u", "/app/main.py"],
                    "env": [
                        {"name": "GAME_TYPE", "value": game_id},
                        {"name": "MANAGER_API_URL", "value": manager_api_url},
                        {"name": "TELEMETRY_API_URL", "value": telemetry_api_url},
                        {"name": "SIDECAR_API_KEY", "value": sidecar_token},
                        {"name": "SERVER_UUID", "value": str(logical_server_id)},
                        {
                            "name": "TARGET_CONTAINER_NAME",
                            "valueFrom": {
                                "fieldRef": {
                                    "fieldPath": "metadata.name"
                                }
                            }
                        }
                    ],
                    "volumeMounts": [
                        {"name": "world-data", "mountPath": "/config", "readOnly": True},
                        {"name": "sidecar-scripts-volume", "mountPath": "/app"}
                    ]
                }
            ],
            "volumes": [
                {"name": "world-data", "persistentVolumeClaim": {"claimName": pvc_name}},
                {"name": "sidecar-scripts-volume", "configMap": {"name": "sidecar-scripts"}}
            ]
        }
        
        # Only add the initContainers key if there are actually containers to run
        if init_containers:
            manifest_spec_template_spec["initContainers"] = init_containers
        
        return {
            "apiVersion": f"{self.group}/{self.version}",
            "kind": "GameServer",
            "metadata": {
                "generateName": f"{game_id}-{mod_state}-{user_id}-",
                "labels": {
                    "craftcloud.role": "game_sidecar",
                    "craftcloud.server_id": str(logical_server_id),
                    "game": game_id,
                    "mod_state": mod_state,
                    "owner_id": str(user_id),
                    "role": "game-sidecar"
                }
            },
            "spec": {
                "container": "game-engine",
                "ports": blueprint.get("ports", []),
                "template": {
                    "spec": manifest_spec_template_spec
                }
            }
        }

    def create_server(self, game_id: str, user_id: str, logical_server_id: str, sidecar_token: str, config_data: dict = None):
        if config_data is None:
            config_data = {}
            
        manifest = self._build_manifest(game_id, user_id, config_data, logical_server_id, sidecar_token)
        
        try:
            response = self.custom_api.create_namespaced_custom_object(
                group=self.group,
                version=self.version,
                namespace=self.namespace,
                plural=self.plural,
                body=manifest
            )
            
            # Return an object that mimics your expected container schema
            class MockContainer:
                def __init__(self, name):
                    self.name = name
                    self.status = "starting"
            
            return MockContainer(name=response["metadata"]["name"])
        except ApiException as e:
            print(f"Kubernetes Deployment Error: {e}")
            raise Exception("Failed to execute Kubernetes deployment.")

    def get_container(self, server_id: str):
        try:
            gs = self.custom_api.get_namespaced_custom_object(
                group=self.group, version=self.version, namespace=self.namespace,
                plural=self.plural, name=server_id
            )
            state = gs.get("status", {}).get("state", "Unknown")
            is_running = state in ["Ready", "Allocated"]
            
            class MockContainer:
                def __init__(self, name, status):
                    self.name = name
                    self.status = "running" if status else "offline"
            return MockContainer(name=server_id, status=is_running)
        except ApiException:
            return None

    def stop_server(self, server_id: str):
        try:
            self.custom_api.delete_namespaced_custom_object(
                group=self.group, version=self.version, namespace=self.namespace,
                plural=self.plural, name=server_id
            )
        except ApiException as e:
            if e.status != 404:
                print(f"Error stopping server: {e}")
    
    def delete_server_and_data(self, server_id: str, logical_server_id: str):
        """Stops the game server and permanently deletes the world data (PVC)."""
        """ Not Yet implemented """
        self.stop_server(server_id)
        
        pvc_name = f"world-pvc-{logical_server_id}"
        try:
            print(f"Permanently deleting storage volume: {pvc_name}")
            self.core_v1.delete_namespaced_persistent_volume_claim(
                name=pvc_name, 
                namespace=self.namespace
            )
        except ApiException as e:
            if e.status == 404:
                print(f"Volume {pvc_name} already deleted or not found.")
            else:
                print(f"Error deleting PVC {pvc_name}: {e}")
                raise e

    def list_all_servers(self):
        server_list = []
        try:
            gameservers = self.custom_api.list_namespaced_custom_object(
                group=self.group, version=self.version, namespace=self.namespace,
                plural=self.plural
            )
            for gs in gameservers.get("items", []):
                metadata = gs.get("metadata", {})
                status = gs.get("status", {})
                current_state = status.get("state", "Unknown")
                
                server_list.append({
                    "id": metadata.get("name"),
                    "name": metadata.get("name").replace("-", " ").title(),
                    "status": "online" if current_state in ["Ready", "Allocated"] else "offline",
                    "cpu": 0, "ram": 0, "players": 0 
                })
        except ApiException as e:
            print(f"Failed to list servers: {e}")
        return server_list

File: services/fleet-manager/internal/k8s/manager.go
package k8s

import (
	"context"
	"fmt"
	"os"
	"path/filepath"

	"fleet-manager/internal/config"

	// Native Kubernetes & Agones Types
	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
	agonesclient "agones.dev/agones/pkg/client/clientset/versioned"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/errors"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)

type ServerManager struct {
	k8sClient    *kubernetes.Clientset
	agonesClient *agonesclient.Clientset
	cfg          *config.Config
}

type ContainerInfo struct {
	Name   string
	Status string
}

func NewServerManager(cfg *config.Config) (*ServerManager, error) {
	k8sConfig, err := rest.InClusterConfig()
	if err != nil {
		homeDir, _ := os.UserHomeDir()
		kubeconfigPath := filepath.Join(homeDir, ".kube", "config")
		k8sConfig, err = clientcmd.BuildConfigFromFlags("", kubeconfigPath)
		if err != nil {
			return nil, fmt.Errorf("failed to load kubeconfig: %v", err)
		}
	}

	// 1. Standard Client (for PVCs)
	k8sClient, err := kubernetes.NewForConfig(k8sConfig)
	if err != nil {
		return nil, err
	}

	// 2. Strongly Typed Agones Client
	agClient, err := agonesclient.NewForConfig(k8sConfig)
	if err != nil {
		return nil, err
	}

	return &ServerManager{
		k8sClient:    k8sClient,
		agonesClient: agClient,
		cfg:          cfg,
	}, nil
}

func (m *ServerManager) CreateServer(gameID, userID, logicalServerID, sidecarToken string, configData map[string]interface{}) (*ContainerInfo, error) {
	ctx := context.Background()

	isModded, _ := configData["is_modded"].(bool)
	modState := map[bool]string{true: "modded", false: "vanilla"}[isModded]

	blueprint := m.cfg.Templates[gameID][modState]

	pvcName, err := m.ensurePVCExists(ctx, logicalServerID)
	if err != nil {
		return nil, fmt.Errorf("failed to provision PVC: %v", err)
	}

	// 1. Build Environment Variables using k8s corev1 types
	var envVars []corev1.EnvVar
	for _, env := range blueprint.EnvDefaults {
		envVars = append(envVars, corev1.EnvVar{Name: env.Name, Value: env.Value})
	}
	for k, v := range configData {
		if k != "is_modded" {
			envVars = append(envVars, corev1.EnvVar{Name: k, Value: fmt.Sprintf("%v", v)})
		}
	}

	// 2. Define Shared Volumes
	volumes := []corev1.Volume{
		{
			Name: "world-data",
			VolumeSource: corev1.VolumeSource{
				PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
					ClaimName: pvcName,
				},
			},
		},
		{
			Name: "sidecar-scripts-volume",
			VolumeSource: corev1.VolumeSource{
				ConfigMap: &corev1.ConfigMapVolumeSource{
					LocalObjectReference: corev1.LocalObjectReference{Name: "sidecar-scripts"},
				},
			},
		},
	}

	// 3. Construct Strongly Typed Agones GameServer
	gameServer := &agonesv1.GameServer{
		ObjectMeta: metav1.ObjectMeta{
			GenerateName: fmt.Sprintf("%s-%s-%s-", gameID, modState, userID),
			Labels: map[string]string{
				"craftcloud.role":      "game_sidecar",
				"craftcloud.server_id": logicalServerID,
				"game":                 gameID,
				"mod_state":            modState,
				"owner_id":             userID,
				"role":                 "game-sidecar",
			},
		},
		Spec: agonesv1.GameServerSpec{
			Container: "game-engine",
			// We map the dynamic ports from YAML into Agones types
			Ports: buildAgonesPorts(blueprint.Ports), 
			Template: corev1.PodTemplateSpec{
				Spec: corev1.PodSpec{
					Volumes: volumes,
					Containers: []corev1.Container{
						{
							Name:  "game-engine",
							Image: blueprint.Image,
							Env:   envVars,
							VolumeMounts: []corev1.VolumeMount{
								{Name: "world-data", MountPath: "/config"},
							},
						},
						{
							Name:    "activity-sidecar",
							Image:   "python:3.14.1-alpine3.23",
							Command: []string{"python", "-u", "/app/main.py"},
							Env: []corev1.EnvVar{
								{Name: "GAME_TYPE", Value: gameID},
								{Name: "MANAGER_API_URL", Value: m.cfg.ManagerAPIURL},
								{Name: "TELEMETRY_API_URL", Value: m.cfg.TelemetryAPIURL},
								{Name: "SIDECAR_API_KEY", Value: sidecarToken},
								{Name: "SERVER_UUID", Value: logicalServerID},
								{
									Name: "TARGET_CONTAINER_NAME",
									ValueFrom: &corev1.EnvVarSource{
										FieldRef: &corev1.ObjectFieldSelector{FieldPath: "metadata.name"},
									},
								},
							},
							VolumeMounts: []corev1.VolumeMount{
								{Name: "world-data", MountPath: "/config", ReadOnly: true},
								{Name: "sidecar-scripts-volume", MountPath: "/app"},
							},
						},
					},
				},
			},
		},
	}

	// 4. Create via Typed Client
	created, err := m.agonesClient.AgonesV1().GameServers(m.cfg.Namespace).Create(ctx, gameServer, metav1.CreateOptions{})
	if err != nil {
		return nil, fmt.Errorf("kubernetes deployment error: %v", err)
	}

	return &ContainerInfo{Name: created.Name, Status: "starting"}, nil
}

func (m *ServerManager) GetContainer(serverID string) *ContainerInfo {
	ctx := context.Background()
	gs, err := m.agonesClient.AgonesV1().GameServers(m.cfg.Namespace).Get(ctx, serverID, metav1.GetOptions{})
	if err != nil {
		return nil
	}

	isRunning := gs.Status.State == agonesv1.GameServerStateReady || gs.Status.State == agonesv1.GameServerStateAllocated
	
	return &ContainerInfo{
		Name:   serverID,
		Status: map[bool]string{true: "running", false: "offline"}[isRunning],
	}
}

func (m *ServerManager) StopServer(serverID string) error {
	ctx := context.Background()
	err := m.agonesClient.AgonesV1().GameServers(m.cfg.Namespace).Delete(ctx, serverID, metav1.DeleteOptions{})
	if err != nil && !errors.IsNotFound(err) {
		return err
	}
	return nil
}

func (m *ServerManager) ensurePVCExists(ctx context.Context, logicalServerID string) (string, error) {
	// (Omitted for brevity - exact same logic as before, using standard k8sClient.CoreV1())
	return fmt.Sprintf("world-pvc-%s", logicalServerID), nil
}

// Helper to cast YAML ports to Agones typed ports
func buildAgonesPorts(yamlPorts []interface{}) []agonesv1.GameServerPort {
	var parsed []agonesv1.GameServerPort
    // ... basic type assertions mapping your interface{} to agonesv1.GameServerPort 
	return parsed
}